collection.go

```diff
@@ -424,6 +438,44 @@ func lazyLoadCollection(coll *CollectionSpec, opts *CollectionOptions) (
 		return prog, nil
 	}
 
+	populateMaps = func() error {
+		for mapName, m := range maps {
+			// TODO: Store reference to MapSpec in Map to avoid this?
+			mapSpec, ok := coll.Maps[mapName]
+			if !ok {
+				return fmt.Errorf("missing map spec %s", mapName)
+			}
+
+			mapSpec = mapSpec.Copy()
+
+			// Replace any object stubs with loaded objects.
+			for i, kv := range mapSpec.Contents {
+				switch v := kv.Value.(type) {
+				case programStub:
+					prog, ok := progs[string(v)]
+					if !ok {
+						return fmt.Errorf("program %s not assigned, but map %s requires it: %w", v, mapName, errUnsatisfiedReference)
```

Comment:
I think this is a bit problematic, because it breaks lazy loading a bit: it's now not enough to request the maps and programs your code needs, you also need to add the transitive dependencies of maps. That isn't straight forward, except in the "load everything always" case.

Maybe its enough to call `loadMap` here? Same for `loadProgram`.

Reply:
Okay, will give it a try. I'm concerned about refcounting traps since we won't be holding on to any references to inner maps/progs. Programs don't hold direct references to inner objects either, so they will be deallocated when the outer map key holding them is overwritten. It feels safer failing explicitly. If we go for the hands-off approach of automatically lazy-loading dependencies, this would need to be well-documented, but users tend to ignore documentation whenever they possibly can.

Reply:
> I'm concerned about refcounting traps since we won't be holding on to any references to inner maps/progs.

Hmm you mean the behaviour that program arrays are removed when the user space handle goes away? Yeah, that is really annoying. I remember that there is a flag which restores a sane behaviour, could we force enable this for ProgramArrays?

Reply:
Program arrays can be pinned, I don't think that's an issue. Let's say userspace creates a program and a program array, writes a reference to the program to prog array index 0, and then closes the fd because it was not requested by `LoadAndAssign`. The last reference to that program is the entry in index 0, and if that map key gets overwritten, the program gets deallocated. I also vaguely remember prog_arrays being write-only, so the program reference can't be retrieved from the prog_array.

> I remember that there is a flag which restores a sane behaviour

Sounds like something we should take into account, do you have more info?

Reply:
>Program arrays can be pinned, I don't think that's an issue. Let's say userspace creates a program and a program array, writes a reference to the program to prog array index 0, and then closes the fd because it was not requested by LoadAndAssign.

I only half remembered the problem, sorry. In your scenario there is no problem when closing the program fd. The issue is that closing the fd for the program array (so, removing the last user space reference) will clear all slots of the program array even though it might still be referenced by the BPF. The result is missed tail calls. This is a really big footgun that has bitten us in production before. I thought that `BPF_F_PRESERVE_ELEMS` would fix this, but that flag only works for perf event arrays, not program arrays.

For map auto population this means that a program array that is populated by the library has to be explicitly requested / held on to by the user. Otherwise tail calls will fail mysteriously.

In the best case we could detect that a program array is only referenced by the BPF and return an error? Unfortunately that requires fudging lazyLoadCollection even more to distinguish between "map required by BPF" and "map required by user".

Otherwise we could ignore the problem.

> The last reference to that program is the entry in index 0, and if that map key gets overwritten, the program gets deallocated.

That's sane behaviour, don't you think?

Reply:
>> The last reference to that program is the entry in index 0, and if that map key gets overwritten, the program gets deallocated.

> That's sane behaviour, don't you think?

Agreed, that would require an explicit action on behalf of the user, and is not something we can account for.

Re: losing prog_array references: this is something we should prevent IMO, but can be addressed outside of this PR since it would require a refactor of the lazy-loader. We should probably merge this one as-is and land a refactor before releasing 0.7.0.

Created https://github.com/cilium/ebpf/issues/366 for this.